### PR TITLE
lazily fetch messages in chat

### PR DESCRIPTION
**Pull request by ryanfkeepers**
once we're past kopia assits, the chat can
download all of its messages and store them in the body uploded to kopia.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #5063 

#### Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/pkg/services/m365/api/teamsChats.go [hunk 1]
```diff
@@ -62,13 +63,47 @@ func (c Chats) GetChatByID(
...
+		lastMsgCreatedAt time.Time
+		members          = chat.GetMembers()
+		memberNames      = []string{}
+		msgs             = chat.GetMessages()
+		preview          string
+		err              error
+	)
+
+	if lastMsgPreview != nil {
+		preview, _, err = getChatMessageContentPreview(lastMsgPreview, noAttachments{})
```
Comment (HiteshRepo) [inline src/pkg/services/m365/api/teamsChats.go:79]:
```suggestion
getChatMessageContentPreview(lastMsgPreview, nil)
```

Comment (HiteshRepo) [inline src/pkg/services/m365/api/teamsChats.go:79]:
```suggestion
getChatMessageContentPreview(lastMsgPreview, nil)
```

Comment (HiteshRepo) [inline src/pkg/services/m365/api/teamsChats.go:79]:
```suggestion
preview, _, err = getChatMessageContentPreview(lastMsgPreview, noAttachments{})
```

File: src/pkg/services/m365/api/channels.go [hunk 2]
```diff
@@ -239,12 +239,28 @@ func GetChatMessageFrom(msg models.ChatMessageable) string {
...
+	GetAttachments() []models.ChatMessageAttachmentable
+}
+
+func getChatMessageContentPreview(msg getBodyer, atts getAttachmentser) (string, int64, error) {
+	content, origSize, err := stripChatMessageHTML(msg, atts)
 	return str.Preview(content, 128), origSize, clues.Stack(err).OrNil()
 }
 
-func stripChatMessageHTML(msg models.ChatMessageable) (string, int64, error) {
+func stripChatMessageHTML(msg getBodyer, atts getAttachmentser) (string, int64, error) {
```
Comment (HiteshRepo) [inline src/pkg/services/m365/api/channels.go:263]:
```suggestion
func stripChatMessageHTML(msg getBodyer, atts []models.ChatMessageAttachmentable
```

File: src/pkg/services/m365/api/channels.go [hunk 3]
```diff
@@ -256,7 +272,7 @@ func stripChatMessageHTML(msg models.ChatMessageable) (string, int64, error) {
 
 	origSize = int64(len(content))
 
-	content = replaceAttachmentMarkup(content, msg.GetAttachments())
+	content = replaceAttachmentMarkup(content, atts.GetAttachments())
```
Comment (HiteshRepo) [inline src/pkg/services/m365/api/channels.go:275]:
```suggestion
	content = replaceAttachmentMarkup(content, atts)
```

File: src/pkg/services/m365/api/channels.go [hunk 4]
```diff
@@ -162,7 +162,7 @@ func channelMessageInfo(
 		modTime = lastReplyAt
 	}
 
-	preview, contentLen, err := getChatMessageContentPreview(msg)
+	preview, contentLen, err := getChatMessageContentPreview(msg, msg)
```
Comment (HiteshRepo) [inline src/pkg/services/m365/api/channels.go:165]:
```suggestion
	preview, contentLen, err := getChatMessageContentPreview(msg, msg.GetAttachments())
```

File: src/pkg/services/m365/api/channels.go [hunk 5]
```diff
@@ -180,7 +180,7 @@ func channelMessageInfo(
 	var lr details.ChannelMessageInfo
 
 	if lastReply != nil {
-		preview, contentLen, err = getChatMessageContentPreview(lastReply)
+		preview, contentLen, err = getChatMessageContentPreview(lastReply, lastReply)
```
Comment (HiteshRepo) [inline src/pkg/services/m365/api/channels.go:183]:
```suggestion
		preview, contentLen, err = getChatMessageContentPreview(lastReply, lastReply.GetAttachments())
```

**Review by HiteshRepo — APPROVED** [review]
Looks good.
I feel we do not need below (hence suggested changes):
```
// a hack for fulfilling getAttachmentser when the model doesn't
// provide GetAttachments()
type noAttachments struct{}

func (noAttachments) GetAttachments() []models.ChatMessageAttachmentable {
	return []models.ChatMessageAttachmentable{}
}

type getAttachmentser interface {
	GetAttachments() []models.ChatMessageAttachmentable
}
```

File: src/pkg/services/m365/api/channels.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/pkg/services/m365/api/channels.go:263]: good idea!

**Review by pandeyabs — APPROVED** [review]

**Pull request closed without merging**